File: nfc.js
// nfc.js
require('dotenv').config();

const { NFC } = require('nfc-pcsc');
const fetch = global.fetch || require('node-fetch');

const API_URL = process.env.API_URL;

const nfc = new NFC();

let lastTagId = null;
let lastScanTime = 0;

console.log("NFC待機中...");

nfc.on('reader', reader => {

    reader.on('card', async card => {
        const now = Date.now();

        // 🔥 連打防止（2秒以内無視）
        if (now - lastScanTime < 2000) {
            return;
        }
        lastScanTime = now;

        const tagId = card.uid;
        console.log("検出:", tagId);

        let sendId;

        // 🔁 トグル
        if (tagId === lastTagId) {
            sendId = "0";
            lastTagId = null;
            console.log("→ OFF");
        } else {
            sendId = tagId;
            lastTagId = tagId;
            console.log("→ ON");
        }

        try {
            await fetch(API_URL, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({
                    tag_id: sendId,
                    timestamp: new Date().toISOString()
                })
            });
        } catch (e) {
            console.error("送信エラー:", e);
        }
    });
});

File: nfc.py
import nfc
import requests
from datetime import datetime
import os
from dotenv import load_dotenv

# .env読み込み
load_dotenv()

API_URL = os.getenv("API_URL")

if not API_URL:
    raise ValueError("API_URLが.envに設定されていません")

last_tag_id = None

def on_connect(tag):
    global last_tag_id

    tag_id = tag.identifier.hex()
    print("検出:", tag_id)

    if tag_id == last_tag_id:
        send_id = "0"
        last_tag_id = None
        print("→ OFF")
    else:
        send_id = tag_id
        last_tag_id = tag_id
        print("→ ON")

    requests.post(API_URL, json={
        "tag_id": send_id,
        "timestamp": datetime.now().isoformat()
    })

    return True

clf = nfc.ContactlessFrontend('usb')

print("NFC待機中...")
clf.connect(rdwr={'on-connect': on_connect})